[BUG]: Getting error while running it

Module not found: Error: Can't resolve 'crypto' in '/Users/patrick/git/patrick/github-activity/node_modules/@type-cacheable/core/dist/util'

BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
        - add a fallback 'resolve.fallback: { "crypto": require.resolve("crypto-browserify") }'
        - install 'crypto-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
        resolve.fallback: { "crypto": false }

--- a/README.md
+++ b/README.md
@@ -51,7 +51,7 @@ cacheManager.setOptions(<CacheManagerOptions>{
 Currently, there are three decorators available in this library: `@Cacheable`, `@CacheClear`, and `@CacheUpdate`. Here is a sample of how they can be used:
 
 ```ts
-import * as Redis from 'redis';
+import Redis from 'redis';
 import { Cacheable, CacheClear } from '@type-cacheable/core';
 
 const userClient = Redis.createClient();

--- a/packages/core/lib/strategies/NoOpStrategy.ts
+++ b/packages/core/lib/strategies/NoOpStrategy.ts
@@ -1,0 +1,28 @@
+import {
+  CacheClearStrategy, CacheClearStrategyContext,
+  CacheStrategy,
+  CacheStrategyContext,
+  CacheUpdateStrategy,
+  CacheUpdateStrategyContext,
+} from '@type-cacheable/core';
+
+export class NoOpStrategy implements CacheStrategy {
+  async handle(context: CacheStrategyContext): Promise<any> {
+    return context.originalMethod!.apply(
+      context.originalMethodScope,
+      context.originalMethodArgs,
+    );
+  }
+}
+
+export class NoOpUpdateStrategy implements CacheUpdateStrategy {
+  async handle(context: CacheUpdateStrategyContext): Promise<any> {
+    return Promise.resolve(undefined);
+  }
+}
+
+export class NoOpClearStrategy implements CacheClearStrategy {
+  async handle(context: CacheClearStrategyContext): Promise<any> {
+    return Promise.resolve(undefined);
+  }
+}

--- a/packages/core/lib/strategies/index.ts
+++ b/packages/core/lib/strategies/index.ts
@@ -1,1 +1,4 @@
+export { DefaultClearStrategy } from './DefaultClearStrategy';
+export { DefaultUpdateStrategy } from './DefaultUpdateStrategy';
 export { DefaultStrategy } from './DefaultStrategy';
+export { NoOpClearStrategy, NoOpUpdateStrategy, NoOpStrategy } from './NoOpStrategy';

--- a/packages/core/lib/util/getCacheKey.ts
+++ b/packages/core/lib/util/getCacheKey.ts
@@ -1,7 +1,6 @@
-import { createHash } from 'crypto';
-import * as serialize from 'serialize-javascript';
-import { CacheKeyBuilder, CacheKeyDeleteBuilder } from '../interfaces';
-import { PostRunKeyBuilder } from '../interfaces/PostRunKeyBuilder';
+import md5 from 'blueimp-md5';
+import serialize from 'serialize-javascript';
+import { CacheKeyBuilder, CacheKeyDeleteBuilder, PostRunKeyBuilder } from '../interfaces';
 
 export type CacheClearKey = string | string[] | CacheKeyDeleteBuilder;
 export type CacheableKey = string | CacheKeyBuilder;
@@ -17,7 +16,7 @@ export type CacheUpdateKey = string | CacheKeyBuilder | PostRunKeyBuilder;
  * @returns {String}
  */
 export const extractKey = (
-  passedInKey: string | string[] | CacheKeyBuilder | CacheKeyDeleteBuilder | PostRunKeyBuilder = '',
+  passedInKey: string | string[] | CacheKeyBuilder | CacheUpdateKey | CacheKeyDeleteBuilder | PostRunKeyBuilder = '',
   args: any[],
   context?: any,
   returnValue?: any,
@@ -40,7 +39,7 @@ export const extractKey = (
  * @returns {String}
  */
 export const getCacheKey = (
-  passedInKey: string | string[] | CacheKeyBuilder | CacheKeyDeleteBuilder | PostRunKeyBuilder = '',
+  passedInKey: string | string[] | CacheKeyBuilder | CacheKeyDeleteBuilder | CacheUpdateKey | PostRunKeyBuilder = '',
   methodName: string,
   args: any[],
   context?: any,
@@ -61,11 +60,11 @@ export const getCacheKey = (
   };
 
   const serializedKey = serialize(callMap);
-  return createHash('md5').update(serializedKey).digest('hex');
+  return md5(serializedKey);
 };
 
 export const getFinalKey = (
-  passedCacheKey: CacheableKey | CacheClearKey | PostRunKeyBuilder = '',
+  passedCacheKey: CacheableKey | CacheClearKey | CacheUpdateKey | PostRunKeyBuilder = '',
   passedHashKey: string | CacheKeyBuilder | PostRunKeyBuilder = '',
   methodName: string,
   args: any[],

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -38,11 +38,13 @@
   },
   "homepage": "https://github.com/joshuaslate/type-cacheable#readme",
   "devDependencies": {
+    "@types/blueimp-md5": "^2.18.0",
     "@types/node": "^20.1.0",
     "@types/serialize-javascript": "^5.0.1",
     "typescript": "^5.0.2"
   },
   "dependencies": {
+    "blueimp-md5": "^2.19.0",
     "reflect-metadata": "^0.1.13",
     "serialize-javascript": "^6.0.0"
   },

--- a/packages/core/tsconfig.json
+++ b/packages/core/tsconfig.json
@@ -9,7 +9,8 @@
     "strict": true,
     "sourceMap": true,
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
   },
   "include": ["lib/**/*"]
 }

--- a/packages/ioredis-adapter/README.md
+++ b/packages/ioredis-adapter/README.md
@@ -23,7 +23,7 @@ yarn add @type-cacheable/core @type-cacheable/ioredis-adapter
 ### Using adapter
 
 ```ts
-import * as IoRedis from 'ioredis';
+import IoRedis from 'ioredis';
 import { useAdapter } from '@type-cacheable/ioredis-adapter';
 
 const client = new IoRedis();

--- a/packages/ioredis-adapter/lib/index.ts
+++ b/packages/ioredis-adapter/lib/index.ts
@@ -1,6 +1,6 @@
 import { Cluster, Redis } from 'ioredis';
 import { compareVersions } from 'compare-versions';
-import cacheManager, { CacheClient, parseIfRequired } from '@type-cacheable/core';
+import cacheManager, { CacheClient, CacheManagerOptions, parseIfRequired } from '@type-cacheable/core';
 
 const REDIS_VERSION_UNLINK_INTRODUCED = '4.0.0';
 const REDIS_VERSION_FRAGMENT_IDENTIFIER = 'redis_version:';
@@ -110,13 +110,17 @@ export class IoRedisAdapter implements CacheClient {
   }
 }
 
-export const useAdapter = (client: Redis, asFallback?: boolean): IoRedisAdapter => {
+export const useAdapter = (client: Redis, asFallback?: boolean, options?: CacheManagerOptions): IoRedisAdapter => {
   const ioRedisAdapter = new IoRedisAdapter(client);
 
   if (asFallback) {
     cacheManager.setFallbackClient(ioRedisAdapter);
   } else {
     cacheManager.setClient(ioRedisAdapter);
+  }
+
+  if (options) {
+    cacheManager.setOptions(options);
   }
 
   return ioRedisAdapter;

--- a/packages/ioredis-adapter/tsconfig.json
+++ b/packages/ioredis-adapter/tsconfig.json
@@ -9,7 +9,8 @@
     "sourceMap": true,
     "strict": true,
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
   },
   "include": ["lib/**/*"]
 }

--- a/packages/lru-cache-adapter/lib/index.ts
+++ b/packages/lru-cache-adapter/lib/index.ts
@@ -1,5 +1,5 @@
 import { LRUCache }  from 'lru-cache';
-import cacheManager, { CacheClient } from '@type-cacheable/core';
+import cacheManager, { CacheClient, CacheManagerOptions } from '@type-cacheable/core';
 
 export class LRUCacheAdapter<T extends {}> implements CacheClient {
   constructor(lruClient: LRUCache<string, T>) {
@@ -78,7 +78,8 @@ export class LRUCacheAdapter<T extends {}> implements CacheClient {
 
 export const useAdapter = <T extends {} = {}>(
   client: LRUCache<string, T>,
-  asFallback?: boolean
+  asFallback?: boolean,
+  options?: CacheManagerOptions,
 ): LRUCacheAdapter<T> => {
   const lruCacheAdapter = new LRUCacheAdapter(client);
 
@@ -87,5 +88,10 @@ export const useAdapter = <T extends {} = {}>(
   } else {
     cacheManager.setClient(lruCacheAdapter);
   }
+
+  if (options) {
+    cacheManager.setOptions(options);
+  }
+
   return lruCacheAdapter;
 };

--- a/packages/lru-cache-adapter/test/LRUCacheAdapter.test.ts
+++ b/packages/lru-cache-adapter/test/LRUCacheAdapter.test.ts
@@ -128,6 +128,7 @@ describe('LRUCacheAdapter Tests', () => {
             client: lruCacheAdapter,
             hashKey: 'user',
             cacheKey: (x) => x[0],
+            ttlSeconds: 30,
           })
           async getId(id: string): Promise<string> {
             mockGetIdImplementation();

--- a/packages/lru-cache-adapter/tsconfig.json
+++ b/packages/lru-cache-adapter/tsconfig.json
@@ -9,7 +9,8 @@
     "sourceMap": true,
     "strict": true,
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
   },
   "include": ["lib/**/*"]
 }

--- a/packages/node-cache-adapter/README.md
+++ b/packages/node-cache-adapter/README.md
@@ -23,7 +23,7 @@ yarn add @type-cacheable/core @type-cacheable/node-cache-adapter
 ### Using adapter
 
 ```ts
-import * as NodeCache from 'node-cache';
+import NodeCache from 'node-cache';
 import { useAdapter } from '@type-cacheable/node-cache-adapter';
 
 const client = new NodeCache();

--- a/packages/node-cache-adapter/lib/index.ts
+++ b/packages/node-cache-adapter/lib/index.ts
@@ -1,5 +1,5 @@
 import * as NodeCache from 'node-cache';
-import cacheManager, { CacheClient } from '@type-cacheable/core';
+import cacheManager, { CacheClient, CacheManagerOptions } from '@type-cacheable/core';
 
 export class NodeCacheAdapter implements CacheClient {
   // The node-cache client
@@ -69,13 +69,17 @@ export class NodeCacheAdapter implements CacheClient {
   }
 }
 
-export const useAdapter = (client: NodeCache, asFallback?: boolean): NodeCacheAdapter => {
+export const useAdapter = (client: NodeCache, asFallback?: boolean, options?: CacheManagerOptions): NodeCacheAdapter => {
   const nodeCacheAdapter = new NodeCacheAdapter(client);
 
   if (asFallback) {
     cacheManager.setFallbackClient(nodeCacheAdapter);
   } else {
     cacheManager.setClient(nodeCacheAdapter);
+  }
+
+  if (options) {
+    cacheManager.setOptions(options);
   }
 
   return nodeCacheAdapter;

--- a/packages/node-cache-adapter/lib/index.ts
+++ b/packages/node-cache-adapter/lib/index.ts
@@ -1,4 +1,4 @@
-import * as NodeCache from 'node-cache';
+import NodeCache from 'node-cache';
 import cacheManager, { CacheClient, CacheManagerOptions } from '@type-cacheable/core';
 
 export class NodeCacheAdapter implements CacheClient {

--- a/packages/node-cache-adapter/test/NodeCacheAdapter.test.ts
+++ b/packages/node-cache-adapter/test/NodeCacheAdapter.test.ts
@@ -1,4 +1,4 @@
-import * as NodeCache from 'node-cache';
+import NodeCache from 'node-cache';
 import { Cacheable, CacheClear } from '@type-cacheable/core';
 import { NodeCacheAdapter, useAdapter } from '../lib';
 

--- a/packages/node-cache-adapter/tsconfig.json
+++ b/packages/node-cache-adapter/tsconfig.json
@@ -9,7 +9,8 @@
     "strict": true,
     "sourceMap": true,
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
   },
   "include": ["lib/**/*"]
 }

--- a/packages/redis-adapter/README.md
+++ b/packages/redis-adapter/README.md
@@ -25,7 +25,7 @@ yarn add @type-cacheable/core @type-cacheable/redis-adapter
 ### Using adapter
 
 ```ts
-import * as Redis from 'redis';
+import Redis from 'redis';
 import { useAdapter } from '@type-cacheable/redis-adapter';
 
 const client = Redis.createClient();

--- a/packages/redis-adapter/lib/index.ts
+++ b/packages/redis-adapter/lib/index.ts
@@ -1,6 +1,6 @@
 import { RedisClientType } from '@redis/client/dist/lib/client';
 import { compareVersions } from 'compare-versions';
-import cacheManager, { CacheClient, parseIfRequired } from '@type-cacheable/core';
+import cacheManager, { CacheClient, CacheManagerOptions, parseIfRequired } from '@type-cacheable/core';
 
 // In order to support scalars in hsets (likely not the intended use, but support has been requested),
 // we need at least one key.
@@ -239,13 +239,17 @@ export class RedisAdapter implements CacheClient {
   }
 }
 
-export const useAdapter = (client: RedisClientType, asFallback?: boolean): RedisAdapter => {
+export const useAdapter = (client: RedisClientType, asFallback?: boolean, options?: CacheManagerOptions): RedisAdapter => {
   const redisAdapter = new RedisAdapter(client);
 
   if (asFallback) {
     cacheManager.setFallbackClient(redisAdapter);
   } else {
     cacheManager.setClient(redisAdapter);
+  }
+
+  if (options) {
+    cacheManager.setOptions(options);
   }
 
   return redisAdapter;

--- a/packages/redis-adapter/tsconfig.json
+++ b/packages/redis-adapter/tsconfig.json
@@ -9,7 +9,8 @@
     "strict": true,
     "sourceMap": true,
     "experimentalDecorators": true,
-    "emitDecoratorMetadata": true
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true
   },
   "include": ["lib/**/*"]
 }

